src/webviews/extension-side/dataviewer/dataViewerDependencyService.node.ts

```diff
@@ -4,132 +4,51 @@
-        } finally {
-            tokenSource.dispose();
-        }
+    constructor(
+        @inject(IApplicationShell) applicationShell: IApplicationShell,
+        @inject(IInstaller) installer: IInstaller,
+        @inject(IPythonExecutionFactory) pythonFactory: IPythonExecutionFactory,
+        @inject(IInterpreterService) interpreterService: IInterpreterService,
+        @inject(IsCodeSpace) isCodeSpace: boolean
+    ) {
+        this.withKernel = new KernelDataViewerDependencyImplementation(applicationShell, isCodeSpace);
+        this.withInterpreter = new InterpreterDataViewerDependencyImplementation(
+            applicationShell,
+            installer,
+            pythonFactory,
+            interpreterService,
+            isCodeSpace
+        );
     }
 
-    private async installMissingDependencies(
-        interpreter: PythonEnvironment,
-        tokenSource: CancellationTokenSource
-    ): Promise<void> {
-        sendTelemetryEvent(Telemetry.PythonModuleInstall, undefined, {
-            action: 'displayed',
-            moduleName: ProductNames.get(Product.pandas)!,
-            pythonEnvType: interpreter?.envType
-        });
-        const selection = this.isCodeSpace
-            ? Common.install()
-            : await this.applicationShell.showErrorMessage(
-                  DataScience.pandasRequiredForViewing(),
-                  { modal: true },
-                  Common.install()
-              );
-
-        // All data science dependencies require an interpreter to be passed in
-        // Default to the active interpreter if no interpreter is available
-        const interpreterToInstallDependenciesInto =
-            interpreter || (await this.interpreterService.getActiveInterpreter());
-
-        if (Cancellation.isCanceled(tokenSource.token)) {
-            return;
-        }
-
-        if (selection === Common.install()) {
-            const cancellationPromise = createPromiseFromCancellation({
-                cancelAction: 'resolve',
-                defaultValue: InstallerResponse.Ignore,
-                token: tokenSource.token
-            });
-            // Always pass a cancellation token to `install`, to ensure it waits until the module is installed.
-            const response = await Promise.race([
-                this.installer.install(Product.pandas, interpreterToInstallDependenciesInto, tokenSource),
-                cancellationPromise
-            ]);
-            if (response === InstallerResponse.Installed) {
-                sendTelemetryEvent(Telemetry.UserInstalledPandas);
-            }
+    async checkAndInstallMissingDependencies(executionEnvironment: IKernel | PythonEnvironment): Promise<void> {
+        // IKernel and PythonEnvironment are only types, so I can't compare prototypes or instances of.
+        if (typeof (executionEnvironment as PythonEnvironment).sysPrefix === 'string') {
+            sendTelemetryEvent(Telemetry.DataViewerUsingInterpreter);
```

Comment:
I think this telemetry should moved into the method `withInterpreter.checkAndInstallMissingDependnecies` , else if this method is called in 3 places tomorrow, we'll need to write the same code in 3 places to send the telemtry.